src/middleware/express/make-middleware.ts

```diff
@@ -0,0 +1,62 @@
+/*!
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import onFinished = require('on-finished');
+import {getOrInjectContext} from '../context';
+// Types-only import.
+import * as express from 'express';
+import { makeHttpRequestData } from './make-http-request';
+
+const CHILD_LOG_NAME_SUFFIX = 'applog';
+
+export interface AnnotatedRequestType<LoggerType> extends express.Request {
+  log: LoggerType;
+}
+
+export function makeMiddleware<LoggerType>(projectId: string,
+  emitRequestLog: (httpRequest: HttpRequest, trace: string) => void,
+  makeChildLogger: (childSuffix: string, trace: string) => LoggerType) {
+  return (req: express.Request, res: express.Response, next: express.NextFunction) => {
+    // TODO(ofrobots): use high-resolution timer.
+    const requestStartMs = Date.now();
```

Comment:
should we use an **hrt** now? we get locked into passing as **ms** once we ship a version that accepts it.
honestly nanos is not super useful when dealing with networks so I'm ok with just ms if you are.

Reply:
i guess we could make it polymorphic later and also accept and array [seconds,nanos]

Reply:
My plan to is to switch it to `performance.now()` which gives high resolution time *and* milliseconds, so this API is future-proof. The only reason I am not doing it now is we need a polyfill because `performance.now()` is not available Node 6.